### PR TITLE
Integrate Java type checker and compiler into frontend

### PR DESCRIPTION
**Pull request by 1001mei**
### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Code quality improvements

### How to test

Input a simple Java program (such as the following) into the code editor and click on the `Run` button:
```
public class Main {
    public static void main(String[] args) {
        System.out.println("Hello world");
    }
}
```
Make sure to choose `Java` as the language option.

### Checklist

- [x] I have tested this code
- [ ] I have updated the documentation

### CODE REVIEW

**Comment by martin-henz** [comment]
Great! Here is an example:

```
class HelloWorld {
    public static void main(String[] args) {
        System.out.println("Hello, World!"); 
    }
}
```

Screenshot 2024-04-10 at 6 51 28 PM

**Review by martin-henz — CHANGES_REQUESTED** [review]

File: src/commons/utils/JavaHelper.ts [hunk 1]
```diff
@@ -129,7 +135,11 @@ export async function javaRun(javaCode: string, context: Context) {
               resolve(
                 context.errors.length
                   ? { status: 'error' }
-                  : { status: 'finished', context, value: '' }
+                  : {
+                      status: 'finished',
+                      context,
+                      value: new ProgramTerminatedSuccessfully()
+                    }
```
[the remark this answers is not in the document]
Comment (bryanlohxz) [reply src/commons/utils/JavaHelper.ts:142]: Done!

File: src/commons/utils/JavaHelper.ts [hunk 2]
```diff
@@ -9,6 +11,33 @@ import DisplayBufferService from './DisplayBufferService';
...
+    return Promise.resolve({ status: 'error' });
+  }
+
+  try {
+    const classFile = compileFromSource(javaCode);
+    compiled = {
+      'Main.class': Buffer.from(new BinaryWriter().generateBinary(classFile)).toString('base64')
+    };
+  } catch (e) {
+    stderr('TypeCheck', e);
```
Comment (martin-henz) [inline src/commons/utils/JavaHelper.ts:37]: Looks strange: Shouldn't this be runtime error instead of TypeCheck?

**Review by martin-henz — CHANGES_REQUESTED** [review]
Found a small issue; otherwise looks good!

File: src/commons/utils/JavaHelper.ts (hunk 2 above)
Comment (bryanlohxz) [reply src/commons/utils/JavaHelper.ts:37]: Will is be ok if I changed this to `"Compile"`? Since it catches errors during compiler time.

Comment (bryanlohxz) [reply src/commons/utils/JavaHelper.ts:37]: The `RuntimeErrors` are caught at the return of the `javaRun` method.

Comment (bryanlohxz) [reply src/commons/utils/JavaHelper.ts:37]: Resolved.

image

**Review by martin-henz — APPROVED** [review]

**Pull request merged**